plugins/builtin/target/azure-vmss/plugin/azure.go

```diff
@@ -96,81 +102,108 @@ func (t *TargetPlugin) scaleOut(ctx context.Context, resourceGroup string, vmSca
 }
 
 // scaleIn drain and delete Scale Set instances to match the Autoscaler has deemed required.
-func (t *TargetPlugin) scaleIn(ctx context.Context, resourceGroup string, vmScaleSet string, num int64, config map[string]string) error {
+func (t *TargetPlugin) scaleIn(ctx context.Context, resourceGroup string, vmScaleSet string, num int64, config map[string]string, vmssMode string) error {
 	// Create a logger for this action to pre-populate useful information we
 	// would like on all log lines.
 	log := t.logger.With("action", "scale_in", "resource_group", resourceGroup, "vmss_name", vmScaleSet)
 
-	// Find instance IDs in the target VMSS and perform pre-scale tasks.
-	pager, err := t.vmssVMs.List(ctx, resourceGroup, vmScaleSet,
-		"startswith(instanceView/statuses/code, 'PowerState') eq true",
-		"instanceView/statuses", "instanceView")
+	// Get all VMs in VMSS
+	log.Debug("getting VMs in VMSS")
+	vms, err := t.getVMSSVMs(ctx, resourceGroup, vmssMode, vmScaleSet)
 	if err != nil {
-		return fmt.Errorf("failed to query VMSS instances: %v", err)
-	}
-
-	remoteIDs := []string{}
-	for pager.NotDone() {
-		for _, vm := range pager.Values() {
-			for _, s := range *vm.VirtualMachineScaleSetVMProperties.InstanceView.Statuses {
-				if strings.HasPrefix(*s.Code, "PowerState/") {
-					if *s.Code == "PowerState/running" {
-						log.Debug("found healthy instance", "id", *vm.ID, "instance_id", *vm.InstanceID)
-						remoteIDs = append(remoteIDs, fmt.Sprintf("%s_%s", vmScaleSet, *vm.InstanceID))
-					} else {
-						log.Debug("skipping instance", "id", *vm.ID, "instance_id", *vm.InstanceID, "code", *s.Code)
-					}
-					break
-				}
-			}
-		}
+		return fmt.Errorf("failed to get VMs in VMSS: %w", err)
+	}
 
-		err := pager.NextWithContext(ctx)
-		if err != nil {
-			return fmt.Errorf("failed to list instances in VMSS: %v", err)
-		}
+	// early exit if no VMs found,
+	// should not happen, but checking.
+	if len(vms) == 0 {
+		return fmt.Errorf("no VMs found in VMSS")
+	}
+
+	// early exit if not enough total VMs found between status check and scale in actions,
+	// should not happen, but checking.
+	if num > int64(len(vms)) {
+		return fmt.Errorf("cannot scale in %d instances, only %d total", num, len(vms))
+	}
+
+	// additional processing needed for flexible mode to get status of VMs
+	// for uniform, populate remoteIDs with VMs
+	var remoteIDs []string
+	log.Debug("getting ready remote IDs for Azure ScaleSet instances")
+	switch vmssMode {
+	case orchestrationModeUniform:
+		remoteIDs = vms
+	case orchestrationModeFlexible:
+		remoteIDs = t.readyFlexibleInstances
```

Comment:
We need the lock here as well because we're reading the value. And because `t.readyFlexibleInstances` is a reference type (a slice, see [an example of that behavior](https://go.dev/play/p/HZVUTBBiBAB)), you'll want to shallow-copy it in the process:


```suggestion
		t.readyFlexibleInstancesLock.Lock()
		remoteIDs = slices.Clone(t.readyFlexibleInstances)
		t.readyFlexibleInstancesLock.Unlock()
```